test/playwright/visual.spec.js

```diff
@@ -0,0 +1,100 @@
+/*
+ * Copyright (C) 2020-2022  Yomichan Authors
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+const path = require('path');
+const {test: base, chromium} = require('@playwright/test');
+const root = path.join(__dirname, '../../');
+
+export const test = base.extend({
+    context: async ({ }, use) => { // eslint-disable-line
+        const pathToExtension = path.join(root, 'ext');
+        const context = await chromium.launchPersistentContext('', {
+            // headless: false,
+            args: [
+                '--headless=new',
+                `--disable-extensions-except=${pathToExtension}`,
+                `--load-extension=${pathToExtension}`
+            ]
+        });
+        await use(context);
+        await context.close();
+    },
+    extensionId: async ({context}, use) => {
+        let [background] = context.serviceWorkers();
+        if (!background) {
+            background = await context.waitForEvent('serviceworker');
+        }
+
+        const extensionId = background.url().split('/')[2];
+        await use(extensionId);
+    }
+});
+const expect = test.expect;
+
+test('visual', async ({page, extensionId}) => {
+    // Install jmdict_english.zip
+    await page.goto(`chrome-extension://${extensionId}/settings.html`);
+    await page.waitForTimeout(500); // wait for welcome.html page to load, so we can then bringToFront() -- otherwise, the screenshot can hang
```

Comment:
No, I don't think playwright can inspect the welcome.html popup because it didn't open it. It's the one that yomitan automatically opens on first install.

Reply:
I removed all uses of timeouts using a couple different tricks to access other pages and other frames and close or inspect them directly.